### PR TITLE
feat(cmake): add native support for testing

### PR DESCRIPTION
**Pull request by kylo252**
This is based on the work done in #18674

use [CTest](https://cmake.org/cmake/help/latest/manual/ctest.1.html) to run lint commands

```sh
ctest --test-dir build --show-only
ctest --test-dir build --print-labels
ctest --test-dir build -R 'lintuncrustify-window*'
ctest --test-dir build -R 'lintc-window*'

```
### Highlights

- lintc (`clint`/`uncrustify`)
	- run lint tests based on patterns
	
		```sh
		ctest --test-dir build -R 'lint.*api*'
		```
		
	- ~~run lint tests for _changed_ files only~~ TBD

- full, customizable, reporting which includes timing information, and the ability to export to JUnit format

  ```sh
  ctest --test-dir build -R lintuncrustify --quiet --output-on-failure
  ctest --test-dir build -R lintpy --extra-verbose
  ```
  
- parallel execution for _all_ of the above

  ```console
  $ ctest --test-dir build -R 'lint' -j`nproc` --progress

  100% tests passed, 0 tests failed out of 695

  Label Time Summary:
  lint-all     = 151.16 sec*proc (695 tests)
  lintcfull    = 146.53 sec*proc (690 tests)

  Total Test time (real) =   8.43 sec
  ```

- integration with [cmake-presets](https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html#test-preset)
	
  ```console
  $ ctest --preset lint
  ```

### Supported tests

- [ ] busted, tracked in https://github.com/neovim/neovim/pull/18674
- [ ] check-single-includes
- [ ] clang-tidy
- [x] commit-lint
- [x] uncrustify
- [x] clint
- [x] stylua
- [x] luacheck
- [x] shellcheck
- [x] flake8

  
Related #19169

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -73,3 +73,6 @@ tags
 
 # vim patches
 /vim-*.patch
+
+# ctest artifcats
+Testing
```
Comment (justinmk) [inline .gitignore:78]: can it go in `build/testing/` or something like that?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,62 @@
+add_subdirectory(includes)
+add_subdirectory(functional/fixtures)
+# TODO:
+# - specify *all* the luarocks requirements for tests (what about ffi?)
+# - use impatient to make up for the (sizeable) overhead when running each test independently
```
Comment (justinmk) [inline test/CMakeLists.txt:5]: Is it required to run each test independently or can we turn that off?

**Comment by justinmk** [comment]
With this change, are we going to be able to delete code in ci/ ? Is ctest output going to be duplicating stuff from the busted testrunner? Maybe the busted testrunner "summary" is no longer needed and should be deleted?

Expected outcomes before merging this:

- build log is less noisy, and definitely not _more_ noisy
- we should be able to delete code that is currently doing what ctest would give us

File: .gitignore (hunk 1 above)
Comment (kylo252) [reply .gitignore:78]: I've still haven't had a chance to look into this, see https://github.com/neovim/neovim/pull/18674#discussion_r906810209

File: src/nvim/CMakeLists.txt [hunk 3]
```diff
@@ -130,6 +130,12 @@ glob_wrapper(UNIT_TEST_FIXTURES ${PROJECT_SOURCE_DIR}/test/unit/fixtures/*.c)
 list(SORT NVIM_SOURCES)
 list(SORT NVIM_HEADERS)
 
+add_library(nvim_sources INTERFACE)
+target_sources(nvim_sources INTERFACE ${NVIM_SOURCES})
+
+add_library(nvim_headers INTERFACE ${NVIM_HEADERS})
+target_sources(nvim_headers INTERFACE ${NVIM_HEADERS})
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt:137]: if this is truly needed, it should be coupled with removing the other code where we are collecting sources/headers and building nvim. We should not have 2 different places where we're doing the same thing in a different way.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+set(NVIM_PRG $<TARGET_FILE:nvim>)
+
+set(test_env)
+list(
+  APPEND test_env
+  "LC_ALL=en_US.UTF-8"
+  "NVIM_PRG=${NVIM_PRG}"
+  "NEOVIM_BUILD_DIR=${PROJECT_BINARY_DIR}"
+  "NEOVIM_SOURCE_DIR=${PROJECT_SOURCE_DIR}"
```
[the remark this answers is not in the document]
Comment (kylo252) [reply test/CMakeLists.txt:26]: most of these are only used for busted tests, and some come from `.github/workflow/env.sh`, so I'll remove them since they're not needed here (yet).

File: src/nvim/CMakeLists.txt (hunk 3 above)
Comment (kylo252) [reply src/nvim/CMakeLists.txt:137]: This is done immediately after these collections are created. This list is currently only used in `src/nvim/CMakeLists.txt` and its _child_ directory `src/nvim/po/CMakeLists.txt`, so unless I missed something, this is the first time that this was needed.

That said, we can use this same "modern" pattern to replace this and other similar usages of directory-based properties with target-based ones
https://github.com/neovim/neovim/blob/3b23f561aef87687495ca683f12943d0cd8eff15/CMakeLists.txt#L666

File: Makefile [hunk 5]
```diff
@@ -137,9 +138,21 @@ helphtml: | nvim build/runtime/doc/tags
 functionaltest functionaltest-lua unittest benchmark: | nvim
 	$(BUILD_TOOL) -C build $@
 
-lintlua lintsh lintpy lintuncrustify lintc lintcfull check-single-includes generated-sources lintcommit lint formatc formatlua format: | build/.ran-cmake
+check-single-includes generated-sources formatc formatlua format: | build/.ran-cmake
 	$(CMAKE_PRG) --build build --target $@
 
+build/CTestTestfile.cmake:
```
Comment (dundargoc) [inline Makefile:144]: Isn't this basically the same as the rule `build/.ran-cmake`?

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+set_tests_properties(
+  lintcommit lintlua-stylua lintlua-luacheck lintpy lintsh PROPERTIES
+  LABELS lint-all
+)
+
+add_custom_target(lintc-changed-only)
+add_custom_command(TARGET lintc-changed-only
+  COMMAND bash -c [==[git diff --name-only --relative=src/nvim HEAD | sed 's@[/.]@-@g' | xargs -n1 ctest --test-dir build -L 'lintc' -R]==]
```
Comment (dundargoc) [inline test/CMakeLists.txt:139]: This assumes linux only.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -82,7 +82,9 @@ jobs:
           build_dir=uncrustify/build
           cmake -S $source_dir -B $build_dir -G Ninja -DCMAKE_BUILD_TYPE=Release
           cmake --build $build_dir
-          mkdir -p $HOME/.cache
+          cmake --install $build_dir --prefix "$HOME/.local"
```
Comment (dundargoc) [inline .github/workflows/ci.yml:85]: Hmm, why is this needed?

File: Makefile (hunk 5 above)
Comment (kylo252) [reply Makefile:144]: In a way yes, yet this is more precise and more specifically I needed this so that those target won't trigger if `build/CTestTestfile.cmake` doesn't exist yet, since `build/.ran-cmake` might already exist already.

File: test/CMakeLists.txt (hunk 6 above)
Comment (kylo252) [reply test/CMakeLists.txt:139]: good point! I guess we need to use cmake to parse the output instead, or make it unix-ish exclusive for now

File: .github/workflows/ci.yml (hunk 7 above)
Comment (kylo252) [reply .github/workflows/ci.yml:85]: I can move this into another PR, it's just a small tweak for caching the uncrustify build more reliably.

Comment (kylo252) [reply .github/workflows/ci.yml:85]: it probably was a separate commit at some point, but rebasing is annoying with multiple commits 🥲

**Comment by dundargoc** [comment]
Rebase

**Comment by dundargoc** [comment]
Hold up. This slows linting by 4 times, and that's with parallelization. Without any parallelization it takes 8 times as long. There's no way we're including this in its current form. Is there something missing or is `ctest` just that much slower than cmake?

**Comment by kylo252** [comment]
> Hold up. This slows linting by 4 times, and that's with parallelization. Without any parallelization it takes 8 times as long.

are you comparing these numbers locally or is it the CI?

> There's no way we're including this in its current form. Is there something missing or is `ctest` just that much slower than cmake?

it seems that spawning many test runners  can be expensive, so we can do things a little bit more efficiently.

we need to decide a few things first:

1. do we need to lint all the files in the CI?
2. do we _want_ to lint all the files in the CI?

**Comment by kylo252** [comment]
> Those questions are irrelevant for this PR.
> The question is whether ctest is worth running. If it's much slower then it obviously isn't.

I think you're missing the point here, CTest can run arbitrary commands. A test to lint all files at once can also be created, if that's faster for the CI, so I need to know where bottleneck is.

Here are some benchmarks from my side: 

```console
# on master
❯ hyperfine 'make lintuncrustify' --prepare 'rm -rf build/touches'
Benchmark 1: make lintuncrustify
  Time (mean ± σ):      4.130 s ±  0.075 s    [User: 50.357 s, System: 3.229 s]
  Range (min … max):    4.005 s …  4.269 s    10 runs

#  runs 379 tests on 4 threads
❯ hyperfine 'ctest --test-dir build -R lintuncrustify -j4'
Benchmark 1: ctest --test-dir build -R lintuncrustify -j4
  Time (mean ± σ):      4.666 s ±  0.083 s    [User: 17.463 s, System: 1.003 s]
  Range (min … max):    4.548 s …  4.817 s    10 runs

# runs 379 tests on 16 threads
❯ hyperfine 'ctest --test-dir build -R lintuncrustify -j16'
Benchmark 1: ctest --test-dir build -R lintuncrustify -j16
  Time (mean ± σ):      3.776 s ±  0.089 s    [User: 51.191 s, System: 1.910 s]
  Range (min … max):    3.595 s …  3.914 s    10 runs
```

**Comment by dundargoc** [comment]
> are you comparing these numbers locally or is it the CI?

Locally

> do we need to lint all the files in the CI?
do we want to lint all the files in the CI?

No, I'm talking specifically about locally.

I compared the time to do `ctest --test-dir build -R 'lint' -j`nproc` --progress` compared to `make lint` on `master`.

**Comment by lewis6991** [comment]
> I think you're missing the point here, CTest can run arbitrary commands. A test to lint all files at once can also be created, if that's faster for the CI, so I need to know where bottleneck is.

If you lint all files in one command then that can't be parallelized. So what's the point exactly? What is the point of ctest?

**Comment by kylo252** [comment]
> No, I'm talking specifically about 
> I compared the time to do `ctest --test-dir build -R 'lint' -j$(nproc) --progress` compared to `make lint` on `master`.

could you test it with only uncrustify? I'm gonna remove clint from here anyway.
**important**: make sure to run `rm -rf build/touches` otherwise `make lint` will only run on a few files.

> Also why remove the `make lint` target lol, that's the most useful one.

that's why this is a draft :')

> If you lint all files in one command then that can't be parallelized. So what's the point exactly? What is the point of ctest?

@lewis6991, it's a saner way to run and define tests in cmake. It also provides an easy way to run tests _per_ file, those single tests can then be used to run on changed files only for example. See the PR description for some examples, and also https://github.com/neovim/neovim/pull/18674#issuecomment-1133932505

keep in mind that ctest can just be a an implementation detail for a cmake target, so you shouldn't even notice it. and I also wouldn't want to merge it if it's slower, so no worries there!

**Comment by dundargoc** [comment]
```
Benchmark 1: make lintuncrustify (master)
  Time (mean ± σ):      1.781 s ±  0.043 s    [User: 33.227 s, System: 1.816 s]
  Range (min … max):    1.691 s …  1.845 s    10 runs

Benchmark 2: ctest --test-dir build -R lintuncrustify -j4 (PR)
  Time (mean ± σ):      2.758 s ±  0.029 s    [User: 10.086 s, System: 0.664 s]
  Range (min … max):    2.708 s …  2.801 s    10 runs

Benchmark 3: ctest --test-dir build -R lintuncrustify -j16 (PR)
  Time (mean ± σ):      1.287 s ±  0.016 s    [User: 17.362 s, System: 1.168 s]
  Range (min … max):    1.262 s …  1.316 s    10 runs

Benchmark 4: ctest --test-dir build -R lintuncrustify -j32 (PR)
  Time (mean ± σ):      1.563 s ±  0.025 s    [User: 36.492 s, System: 2.437 s]
  Range (min … max):    1.527 s …  1.614 s    10 runs
```

OK, it seems with only the uncrustify test using `ctest` is faster when using more than 8 jobs on my computer. This makes me even more curious why `make lint` took so much longer. It's not possible to do a direct comparison unless you re-introduce `make lint`.

File: .gitignore (hunk 1 above)
Comment (dundargoc) [reply .gitignore:79]: I don't really understand what you mean? If this isn't needed for this PR then it shouldn't really be here.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (dundargoc) [reply .github/workflows/ci.yml:85]: You did it again. Either learn git rebasing or create a new PR for this lol.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -103,21 +105,22 @@ jobs:
           args: --check runtime/
 
       - if: "!cancelled()"
-        name: lintlua
-        run: make lintlua
+        name: luacheck
+        run: ctest --test-dir ${BUILD_DIR} -R luacheck --output-junit lintluacheck.xml
```
Comment (dundargoc) [inline .github/workflows/ci.yml:109]: I don't understand the `--output-junit` flags. Are we even using `lintluacheck.xml` somewhere?

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+add_test(
+  NAME lintcommit
+  COMMAND ${NVIM_PRG} -u NONE --headless -es -c [==[lua require('scripts.lintcommit').main({trace=false})]==]
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+get_target_property(NVIM_SOURCES nvim_sources INTERFACE_SOURCES)
+get_target_property(NVIM_HEADERS nvim_headers INTERFACE_SOURCES)
```
Comment (dundargoc) [inline test/CMakeLists.txt:27]: OOF, this one hurts bad. These variables were relatively self-contained before, now we're defining these in one subdirectory and using it in another, which kinda breaks the purpose of subdirectories IMO. We might need to rethink this.

Do we really need to move these to `test/`? I mean it makes sense that tests are located in `test/`, but maybe we should just replace the lint targets with tests on their current location?

File: .github/workflows/ci.yml (hunk 8 above)
Comment (kylo252) [reply .github/workflows/ci.yml:109]: I was hoping to get a nice report using a github action, see: https://github.com/neovim/neovim/commit/d13ab1daabdb5411487c84f567260f8289d3ca84

but it seems that it can be tricky here because this repo is a fork 😞 

https://github.com/EnricoMi/publish-unit-test-result-action#support-fork-repositories-and-dependabot-branches

JUnit is a very popular format otherwise, so there's bound to be something that works, but I need help for that https://github.com/marketplace?type=actions&query=junit

Comment (dundargoc) [reply .github/workflows/ci.yml:109]: Can't we do it in a follow-up PR?

File: test/CMakeLists.txt (hunk 9 above)
Comment (kylo252) [reply test/CMakeLists.txt:27]: we can create a function that will add a bunch of tests for a list of source files, what do you think?

File: test/CMakeLists.txt (hunk 6 above)
Comment (kylo252) [reply test/CMakeLists.txt:139]: actually, we should rework/(maybe split?) `cmake/Format.cmake`  so we that `get_changed_files` can be used here
https://github.com/neovim/neovim/blob/6da2271a04540f6f3bd6d3cae1895bb1ea1b5011/cmake/Format.cmake#L3-L4

Comment (kylo252) [reply test/CMakeLists.txt:139]: actually, we should rework/(maybe split?) `cmake/Format.cmake`  so that `get_changed_files` can be used here
https://github.com/neovim/neovim/blob/6da2271a04540f6f3bd6d3cae1895bb1ea1b5011/cmake/Format.cmake#L3-L4

File: test/CMakeLists.txt (hunk 9 above)
Comment (dundargoc) [reply test/CMakeLists.txt:27]: Could work? I dunno, feels easier to just use `add_test` IMO.

File: test/CMakeLists.txt (hunk 6 above)
Comment (dundargoc) [reply test/CMakeLists.txt:139]: Yeah, makes sense. Maybe we should skip `lintc-changed-only`, and focus on not losing any existing functionality before we start worrying about fancy stuff like this.

File: .github/workflows/ci.yml (hunk 8 above)
Comment (kylo252) [reply .github/workflows/ci.yml:109]: yeah that was plan which is why that workflow step isn't here anymore, but I forgot to revert this flag as well

File: test/CMakeLists.txt (hunk 9 above)
Comment (dundargoc) [reply test/CMakeLists.txt:27]: Eh, sure, a function is fine if that's easier.

File: test/CMakeLists.txt (hunk 6 above)
Comment (kylo252) [reply test/CMakeLists.txt:139]: I figured it'd be a slight regression not to add this, since "touch" files aren't needed/considered anymore.

Overall, I vote for skipping this internal integration for now as well, since running all the lint tests is still extremely fast (takes <10s on my machine)

File: .github/workflows/ci.yml (hunk 8 above)
Comment (dundargoc) [reply .github/workflows/ci.yml:109]: OK, just remove it then.

File: test/CMakeLists.txt (hunk 6 above)
Comment (dundargoc) [reply test/CMakeLists.txt:139]: Remove it

File: Makefile (hunk 5 above)
Comment (dundargoc) [reply Makefile:137]: I'm pretty sure this isn't correct. Running

```
make distclean
make lintlua
```

will result in an error because the dependencies aren't built.

File: .gitignore [hunk 10]
```diff
@@ -75,3 +75,5 @@ tags
 /vim-*.patch
 
 /CMakeUserPresets.json
+# ctest artifacts
+Testing
```
Comment (dundargoc) [inline .gitignore:79]: Still what is this?

File: Makefile [hunk 11]
```diff
@@ -130,9 +131,24 @@ src/nvim/testdir/%.vim: phony_force
...
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -R $@
+
+lint: build/CTestTestfile.cmake
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -L 'lint-all'
+
+lintc: build/CTestTestfile.cmake check-single-includes
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -R lintc-changed-only
+
+lintcfull: build/CTestTestfile.cmake
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -L 'lintcfull'
```
Comment (dundargoc) [inline Makefile:150]: Shouldn't these have `-j` to parallelize them?

File: Makefile [hunk 12]
```diff
@@ -130,9 +131,24 @@ src/nvim/testdir/%.vim: phony_force
...
+	$(CMAKE_PRG) -S . -B build
+
+lintlua lintsh lintpy lintcommit: build/CTestTestfile.cmake
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -R $@
+
+lint: build/CTestTestfile.cmake
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -L 'lint-all'
+
+lintc: build/CTestTestfile.cmake check-single-includes
+	ctest --test-dir build $(CTEST_EXTRA_FLAG) -R lintc-changed-only
```
Comment (dundargoc) [inline Makefile:147]: This target doesn't exist anymore

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -103,16 +103,16 @@ jobs:
           args: --check runtime/
 
       - if: "!cancelled()"
-        name: lintlua
-        run: make lintlua
+        name: luacheck
+        run: ctest --test-dir ${BUILD_DIR} -R luacheck
```
Comment (dundargoc) [inline .github/workflows/ci.yml:107]: Just `make lintlua` should work here, right?

File: CMakePresets.json [hunk 14]
```diff
@@ -91,5 +91,30 @@
         "rhs": "Windows"
       }
     }
+  ],
```
Comment (dundargoc) [inline CMakePresets.json:94]: These last second additions really drag out the review process, because now I'll have to check what this is, check that it works and determine if it's worth it. You can leave this as it seems to work, but just keep in mind that if it's not essential then we can iterate on this in a followup PR.

File: .github/workflows/ci.yml (hunk 13 above)
Comment (kylo252) [reply .github/workflows/ci.yml:107]: It does indeed work. 

I was hoping to find a github action that can interact with ctest directly, but we can skip that for now.

File: CMakePresets.json (hunk 14 above)
Comment (kylo252) [reply CMakePresets.json:94]: I'm aware of that, but this one is a pretty useful addition.

File: Makefile (hunk 11 above)
Comment (kylo252) [reply Makefile:150]: oop, I forgot to set [`$CTEST_PARALLEL_LEVEL`](https://cmake.org/cmake/help/latest/envvar/CTEST_PARALLEL_LEVEL.html)

File: .gitignore (hunk 10 above)
Comment (kylo252) [reply .gitignore:79]: this will get generated if you mess up and call `ctest` directly without giving it a test-dir (or a preset)

```sh
cd <project-root>
rm -rf Testing
ctest --show-only # there are probably a few different flags that does this as well
test -d Testing && echo "ctest can be stupid"
```

Comment (dundargoc) [reply .gitignore:79]: Eh, not convincing, if `ctest` generates this it should just be removed instead of being ignored. I'd say skip this.

Comment (kylo252) [reply .gitignore:79]: removed by who? it will just be an annoying folder that gets in the way, same as any other cache folder

see: https://github.com/Kitware/CMake/blob/a3252ac53eeb9f0663a9410e3d48340741c348ea/.gitignore#L12

Comment (kylo252) [reply .gitignore:79]: also see: https://github.com/neovim/neovim/pull/18674#discussion_r906829680

Comment (dundargoc) [reply .gitignore:79]: But it's not a cache folder, you said it shows up if you use the wrong command like `ctest --show-only`. If I accidentally write `mkdir random_dir` then the correct solution isn't to add `random_dir` to `.gitignore`, it's to not write wrong commands. I don't see your point here.

Comment (kylo252) [reply .gitignore:79]: > But it's not a cache folder, you said it shows up if you use the wrong command like `ctest --show-only`.

it's still a cache folder, just in the wrong directory. when used correctly it shows up in test-dir, in our case it's under `build/`

> If I accidentally write `mkdir random_dir` then the correct solution isn't to add `random_dir` to `.gitignore`, it's to not write wrong commands.

Your example doesn't apply here.

- we're adding a new tool that is known to generate a folder that should not be tracked
- this folder has a very _specific_ name, it's always _Testing_
- this makes it a perfect candidate to be added to `.gitignore`
- it will help users realize that this folder is safe to ignore/delete as well
- It's not worth it to try to figure out why and how `ctest` has this bug/feature/behavior
- no file/folder called _Testing_ is currently tracked, and I doubt there will ever be one

so all in all this ranks an easy 9/10 on my scale for QoL/effort

File: Makefile (hunk 5 above)
Comment (kylo252) [reply Makefile:138]: I completely missed your comment, sorry! 

Github isn't showing this [comment](https://github.com/neovim/neovim/pull/19189#discussion_r1008667555) as a new activity/review at the bottom of the page :disappointed:

**Comment by dundargoc** [comment]
I don't think there's much to be gained here unfortunately. The linting is already done in parallel with ninja, and I wouldn't be shocked if its parallelization algorithm is smarter than us slapping on a `-jN`. We don't really need reporting or timing for simple linting either. I'm opting to close this as I believe it's a dead end. Thank you for the effort though.
